File: src/common/wasm/zig/reduce_prod.zig
//! WASM reduction product kernels for all numeric types.
//!
//! Reduction: result = product(a[0..N])
//! uint types reuse signed kernels (wrapping multiplication gives same bits).

const simd = @import("simd.zig");

/// Returns the product of f64 elements. Uses 2-wide SIMD accumulation.
export fn reduce_prod_f64(a: [*]const f64, N: u32) f64 {
    var acc: simd.V2f64 = .{ 1, 1 };
    const n_simd = N & ~@as(u32, 1);
    var i: u32 = 0;
    while (i < n_simd) : (i += 2) {
        acc *= simd.load2_f64(a, i);
    }
    var total: f64 = acc[0] * acc[1];
    while (i < N) : (i += 1) {
        total *= a[i];
    }
    return total;
}

/// Returns the product of f32 elements. Uses 4-wide SIMD accumulation.
export fn reduce_prod_f32(a: [*]const f32, N: u32) f32 {
    var acc: simd.V4f32 = .{ 1, 1, 1, 1 };
    const n_simd = N & ~@as(u32, 3);
    var i: u32 = 0;
    while (i < n_simd) : (i += 4) {
        acc *= simd.load4_f32(a, i);
    }
    var total: f32 = acc[0] * acc[1] * acc[2] * acc[3];
    while (i < N) : (i += 1) {
        total *= a[i];
    }
    return total;
}

/// Returns the product of i64 elements. Uses 2-wide SIMD accumulation.
/// Note: i64x2.mul is native on x86 but emulated on ARM NEON.
/// Handles both signed (i64) and unsigned (u64) since wrapping multiplication gives same bits.
export fn reduce_prod_i64(a: [*]const i64, N: u32) i64 {
    var acc: simd.V2i64 = .{ 1, 1 };
    const n_simd = N & ~@as(u32, 1);
    var i: u32 = 0;
    while (i < n_simd) : (i += 2) {
        acc *%= simd.load2_i64(a, i);
    }
    var total: i64 = acc[0] *% acc[1];
    while (i < N) : (i += 1) {
        total *%= a[i];
    }
    return total;
}

/// Returns the product of i32 elements. Uses 4-wide SIMD accumulation.
/// Handles both signed (i32) and unsigned (u32) since wrapping multiplication gives same bits.
export fn reduce_prod_i32(a: [*]const i32, N: u32) i32 {
    var acc: simd.V4i32 = .{ 1, 1, 1, 1 };
    const n_simd = N & ~@as(u32, 3);
    var i: u32 = 0;
    while (i < n_simd) : (i += 4) {
        acc *%= simd.load4_i32(a, i);
    }
    var total: i32 = acc[0] *% acc[1] *% acc[2] *% acc[3];
    while (i < N) : (i += 1) {
        total *%= a[i];
    }
    return total;
}

/// Returns the product of i16 elements. Promotes to i64 via 2-wide SIMD accumulation.
export fn reduce_prod_i16(a: [*]const i16, N: u32) i64 {
    var acc: simd.V2i64 = .{ 1, 1 };
    const n_simd = N & ~@as(u32, 1);
    var i: u32 = 0;
    while (i < n_simd) : (i += 2) {
        // Widen i16 → i64 and multiply
        acc *%= simd.V2i64{ @as(i64, a[i]), @as(i64, a[i + 1]) };
    }
    var total: i64 = acc[0] *% acc[1];
    while (i < N) : (i += 1) {
        total *%= @as(i64, a[i]);
    }
    return total;
}

/// Returns the product of u16 elements. Promotes to u64 via 2-wide SIMD accumulation.
export fn reduce_prod_u16(a: [*]const u16, N: u32) u64 {
    var acc: simd.V2u64 = .{ 1, 1 };
    const n_simd = N & ~@as(u32, 1);
    var i: u32 = 0;
    while (i < n_simd) : (i += 2) {
        acc *%= simd.V2u64{ @as(u64, a[i]), @as(u64, a[i + 1]) };
    }
    var total: u64 = acc[0] *% acc[1];
    while (i < N) : (i += 1) {
        total *%= @as(u64, a[i]);
    }
    return total;
}

/// Returns the product of i8 elements. Promotes to i64 via 2-wide SIMD accumulation.
export fn reduce_prod_i8(a: [*]const i8, N: u32) i64 {
    var acc: simd.V2i64 = .{ 1, 1 };
    const n_simd = N & ~@as(u32, 1);
    var i: u32 = 0;
    while (i < n_simd) : (i += 2) {
        acc *%= simd.V2i64{ @as(i64, a[i]), @as(i64, a[i + 1]) };
    }
    var total: i64 = acc[0] *% acc[1];
    while (i < N) : (i += 1) {
        total *%= @as(i64, a[i]);
    }
    return total;
}

/// Returns the product of u8 elements. Promotes to u64 via 2-wide SIMD accumulation.
export fn reduce_prod_u8(a: [*]const u8, N: u32) u64 {
    var acc: simd.V2u64 = .{ 1, 1 };
    const n_simd = N & ~@as(u32, 1);
    var i: u32 = 0;
    while (i < n_simd) : (i += 2) {
        acc *%= simd.V2u64{ @as(u64, a[i]), @as(u64, a[i + 1]) };
    }
    var total: u64 = acc[0] *% acc[1];
    while (i < N) : (i += 1) {
        total *%= @as(u64, a[i]);
    }
    return total;
}

/// Returns the product of u64 elements in a strided layout.
export fn reduce_prod_strided_f64(a: [*]const f64, out: [*]f64, outer: u32, axis: u32, inner: u32) void {
    const O = @as(usize, outer);
    const A = @as(usize, axis);
    const I = @as(usize, inner);
    const S = A * I;
    for (0..O) |o| {
        const base = o * S;
        const ob = o * I;
        for (0..I) |i| out[ob + i] = a[base + i];
        for (1..A) |ax| {
            for (0..I) |i| out[ob + i] *= a[base + ax * I + i];
        }
    }
}

/// Returns the product of f32 elements in a strided layout.
export fn reduce_prod_strided_f32(a: [*]const f32, out: [*]f32, outer: u32, axis: u32, inner: u32) void {
    const O = @as(usize, outer);
    const A = @as(usize, axis);
    const I = @as(usize, inner);
    const S = A * I;
    for (0..O) |o| {
        const base = o * S;
        const ob = o * I;
        for (0..I) |i| out[ob + i] = a[base + i];
        for (1..A) |ax| {
            for (0..I) |i| out[ob + i] *= a[base + ax * I + i];
        }
    }
}

/// Returns the product of i64 elements in a strided layout.
export fn reduce_prod_strided_i64(a: [*]const i64, out: [*]i64, outer: u32, axis: u32, inner: u32) void {
    const O = @as(usize, outer);
    const A = @as(usize, axis);
    const I = @as(usize, inner);
    const S = A * I;
    for (0..O) |o| {
        const base = o * S;
        const ob = o * I;
        for (0..I) |i| out[ob + i] = a[base + i];
        for (1..A) |ax| {
            for (0..I) |i| out[ob + i] *%= a[base + ax * I + i];
        }
    }
}

/// Returns the product of u64 elements in a strided layout.
export fn reduce_prod_strided_u64(a: [*]const u64, out: [*]u64, outer: u32, axis: u32, inner: u32) void {
    const O = @as(usize, outer);
    const A = @as(usize, axis);
    const I = @as(usize, inner);
    const S = A * I;
    for (0..O) |o| {
        const base = o * S;
        const ob = o * I;
        for (0..I) |i| out[ob + i] = a[base + i];
        for (1..A) |ax| {
            for (0..I) |i| out[ob + i] *%= a[base + ax * I + i];
        }
    }
}

/// Returns the product of i32 elements in a strided layout. Uses 4-wide SIMD accumulation.
export fn reduce_prod_strided_i32(a: [*]const i32, out: [*]i32, outer: u32, axis: u32, inner: u32) void {
    const O = @as(usize, outer);
    const A = @as(usize, axis);
    const I = @as(usize, inner);
    const S = A * I;
    for (0..O) |o| {
        const base = o * S;
        const ob = o * I;
        for (0..I) |i| out[ob + i] = a[base + i];
        for (1..A) |ax| {
            for (0..I) |i| out[ob + i] *%= a[base + ax * I + i];
        }
    }
}

/// Returns the product of u32 elements in a strided layout.
export fn reduce_prod_strided_u32(a: [*]const u32, out: [*]u32, outer: u32, axis: u32, inner: u32) void {
    const O = @as(usize, outer);
    const A = @as(usize, axis);
    const I = @as(usize, inner);
    const S = A * I;
    for (0..O) |o| {
        const base = o * S;
        const ob = o * I;
        for (0..I) |i| out[ob + i] = a[base + i];
        for (1..A) |ax| {
            for (0..I) |i| out[ob + i] *%= a[base + ax * I + i];
        }
    }
}

/// Returns the product of i16 elements in a strided layout.
/// Promotes to i64 output to avoid overflow.
export fn reduce_prod_strided_i16(a: [*]const i16, out: [*]i64, outer: u32, axis: u32, inner: u32) void {
    const O = @as(usize, outer);
    const A = @as(usize, axis);
    const I = @as(usize, inner);
    const S = A * I;
    for (0..O) |o| {
        const base = o * S;
        const ob = o * I;
        for (0..I) |i| out[ob + i] = @as(i64, a[base + i]);
        for (1..A) |ax| {
            for (0..I) |i| out[ob + i] *%= @as(i64, a[base + ax * I + i]);
        }
    }
}

/// Returns the product of u16 elements in a strided layout.
/// Promotes to u64 output to avoid overflow.
export fn reduce_prod_strided_u16(a: [*]const u16, out: [*]u64, outer: u32, axis: u32, inner: u32) void {
    const O = @as(usize, outer);
    const A = @as(usize, axis);
    const I = @as(usize, inner);
    const S = A * I;
    for (0..O) |o| {
        const base = o * S;
        const ob = o * I;
        for (0..I) |i| out[ob + i] = @as(u64, a[base + i]);
        for (1..A) |ax| {
            for (0..I) |i| out[ob + i] *%= @as(u64, a[base + ax * I + i]);
        }
    }
}

/// Returns the product of i8 elements in a strided layout.
/// Promotes to i64 output to avoid overflow.
export fn reduce_prod_strided_i8(a: [*]const i8, out: [*]i64, outer: u32, axis: u32, inner: u32) void {
    const O = @as(usize, outer);
    const A = @as(usize, axis);
    const I = @as(usize, inner);
    const S = A * I;
    for (0..O) |o| {
        const base = o * S;
        const ob = o * I;
        for (0..I) |i| out[ob + i] = @as(i64, a[base + i]);
        for (1..A) |ax| {
            for (0..I) |i| out[ob + i] *%= @as(i64, a[base + ax * I + i]);
        }
    }
}

/// Returns to product of u8 elements in a strided layout.
/// Promotes to u64 output to avoid overflow.
export fn reduce_prod_strided_u8(a: [*]const u8, out: [*]u64, outer: u32, axis: u32, inner: u32) void {
    const O = @as(usize, outer);
    const A = @as(usize, axis);
    const I = @as(usize, inner);
    const S = A * I;
    for (0..O) |o| {
        const base = o * S;
        const ob = o * I;
        for (0..I) |i| out[ob + i] = @as(u64, a[base + i]);
        for (1..A) |ax| {
            for (0..I) |i| out[ob + i] *%= @as(u64, a[base + ax * I + i]);
        }
    }
}

// --- Tests ---

test "reduce_prod_f64 basic" {
    const testing = @import("std").testing;
    const a = [_]f64{ 1.0, 2.0, 3.0, 4.0, 5.0 };
    try testing.expectApproxEqAbs(reduce_prod_f64(&a, 5), 120.0, 1e-10);
}

test "reduce_prod_i32 basic" {
    const testing = @import("std").testing;
    const a = [_]i32{ 1, 2, 3, 4, 5 };
    try testing.expectEqual(reduce_prod_i32(&a, 5), 120);
}

test "reduce_prod_f64 single and empty" {
    const testing = @import("std").testing;
    const a = [_]f64{7.0};
    try testing.expectApproxEqAbs(reduce_prod_f64(&a, 1), 7.0, 1e-10);
    const b = [_]f64{};
    try testing.expectApproxEqAbs(reduce_prod_f64(&b, 0), 1.0, 1e-10);
}

test "reduce_prod_f64 contains zero" {
    const testing = @import("std").testing;
    const a = [_]f64{ 1.0, 2.0, 0.0, 4.0 };
    try testing.expectApproxEqAbs(reduce_prod_f64(&a, 4), 0.0, 1e-10);
}

test "reduce_prod_f64 negatives" {
    const testing = @import("std").testing;
    const a = [_]f64{ -2.0, 3.0, -4.0 };
    try testing.expectApproxEqAbs(reduce_prod_f64(&a, 3), 24.0, 1e-10);
}

test "reduce_prod_f32 simd boundary" {
    // 4-wide SIMD: N=4, N=5
    const testing = @import("std").testing;
    const a = [_]f32{ 1.0, 2.0, 3.0, 4.0, 5.0 };
    try testing.expectApproxEqAbs(reduce_prod_f32(&a, 4), 24.0, 1e-5);
    try testing.expectApproxEqAbs(reduce_prod_f32(&a, 5), 120.0, 1e-5);
}

test "reduce_prod_i64 basic" {
    const testing = @import("std").testing;
    const a = [_]i64{ 2, 3, 4 };
    try testing.expectEqual(reduce_prod_i64(&a, 3), 24);
}

test "reduce_prod_i64 negatives" {
    const testing = @import("std").testing;
    const a = [_]i64{ -2, -3 };
    try testing.expectEqual(reduce_prod_i64(&a, 2), 6);
}

test "reduce_prod_i16 promotion" {
    // i16 → i64 output to avoid overflow
    const testing = @import("std").testing;
    const a = [_]i16{ 10, 20, 30 };
    try testing.expectEqual(reduce_prod_i16(&a, 3), 6000);
}

test "reduce_prod_i8 promotion" {
    // i8 → i64 output to avoid overflow
    const testing = @import("std").testing;
    const a = [_]i8{ 2, 3, 4, 5 };
    try testing.expectEqual(reduce_prod_i8(&a, 4), 120);
}

test "reduce_prod_u16 promotion" {
    const testing = @import("std").testing;
    const a = [_]u16{ 10, 20, 30 };
    try testing.expectEqual(reduce_prod_u16(&a, 3), 6000);
}

test "reduce_prod_u8 promotion" {
    const testing = @import("std").testing;
    const a = [_]u8{ 2, 3, 4, 5 };
    try testing.expectEqual(reduce_prod_u8(&a, 4), 120);
}

File: src/common/wasm/reduce_prod.ts
/**
 * WASM-accelerated reduction product.
 *
 * Reduction: result = product(a[0..N])
 * Returns null if WASM can't handle this case.
 * uint types route to signed kernels (wrapping multiplication gives same bits).
 */

import {
  reduce_prod_f64,
  reduce_prod_f32,
  reduce_prod_i64,
  reduce_prod_i32,
  reduce_prod_i16,
  reduce_prod_i8,
  reduce_prod_u16,
  reduce_prod_u8,
  reduce_prod_strided_f64,
  reduce_prod_strided_f32,
  reduce_prod_strided_i64,
  reduce_prod_strided_i32,
  reduce_prod_strided_i16,
  reduce_prod_strided_i8,
  reduce_prod_strided_u64,
  reduce_prod_strided_u32,
  reduce_prod_strided_u16,
  reduce_prod_strided_u8,
} from './bins/reduce_prod.wasm';
import {
  resetScratchAllocator,
  resolveInputPtr,
  f16InputToScratchF32,
  alloc,
  copyOut,
} from './runtime';
import { ArrayStorage } from '../storage';
import type { DType, TypedArray } from '../dtype';
import { wasmConfig } from './config';

const BASE_THRESHOLD = 64;

type ReduceFn = (aPtr: number, N: number) => number | bigint;

const kernels: Partial<Record<DType, ReduceFn>> = {
  float64: reduce_prod_f64,
  float32: reduce_prod_f32,
  float16: reduce_prod_f32,
  int64: reduce_prod_i64,
  uint64: reduce_prod_i64,
  int32: reduce_prod_i32,
  uint32: reduce_prod_i32,
  // int8/int16 → i64, uint8/uint16 → u64 (promoted to avoid overflow)
  int16: reduce_prod_i16,
  uint16: reduce_prod_u16,
  int8: reduce_prod_i8,
  uint8: reduce_prod_u8,
};

type AnyTypedArrayCtor = new (length: number) => TypedArray;
const ctorMap: Partial<Record<DType, AnyTypedArrayCtor>> = {
  float64: Float64Array,
  float32: Float32Array,
  float16: Float32Array,
  int64: BigInt64Array,
  uint64: BigUint64Array,
  int32: Int32Array,
  uint32: Uint32Array,
  int16: Int16Array,
  uint16: Uint16Array,
  int8: Int8Array,
  uint8: Uint8Array,
};

/**
 * WASM-accelerated reduction product (no axis, full array).
 * Returns null if WASM can't handle (complex types, non-contiguous, too small).
 */
export function wasmReduceProd(a: ArrayStorage): number | null {
  if (!a.isCContiguous) return null;

  const size = a.size;
  if (size < BASE_THRESHOLD * wasmConfig.thresholdMultiplier) return null;

  const dtype = a.dtype;
  const kernel = kernels[dtype];
  const Ctor = ctorMap[dtype];
  if (!kernel || !Ctor) return null;

  wasmConfig.wasmCallCount++;
  resetScratchAllocator();

  // Float16: convert to f32, run f32 kernel, cast result back to f16 precision.
  // f32 accumulation + f16 cast matches NumPy's pairwise summation behavior.
  if (dtype === 'float16') {
    const aPtr = f16InputToScratchF32(a, size);
    const f32Result = Number(kernel(aPtr, size));
    // Cast through Float16Array to match f16 overflow (e.g., 65536 → inf)
    const f16Round = new Float16Array(1);
    f16Round[0] = f32Result;
    return f16Round[0]!;
  }

  const bpe = (Ctor as unknown as { BYTES_PER_ELEMENT: number }).BYTES_PER_ELEMENT;
  const aPtr = resolveInputPtr(a.data, a.isWasmBacked, a.wasmPtr, a.offset, size, bpe);

  return Number(kernel(aPtr, size));
}

// --- Strided axis reduction for prod ---
// Output type depends on input: narrow ints promote, others stay native.

type StridedFn = (aPtr: number, outPtr: number, outer: number, axis: number, inner: number) => void;

const stridedKernels: Partial<Record<DType, StridedFn>> = {
  float64: reduce_prod_strided_f64,
  float32: reduce_prod_strided_f32,
  float16: reduce_prod_strided_f32,
  int64: reduce_prod_strided_i64,
  uint64: reduce_prod_strided_u64,
  int32: reduce_prod_strided_i32,
  uint32: reduce_prod_strided_u32,
  int16: reduce_prod_strided_i16,
  uint16: reduce_prod_strided_u16,
  int8: reduce_prod_strided_i8,
  uint8: reduce_prod_strided_u8,
};

// Output dtype: narrow ints promote to i64/u64, others stay native
const stridedOutDtype: Partial<Record<DType, DType>> = {
  float64: 'float64',
  float32: 'float32',
  float16: 'float16',
  int64: 'int64',
  uint64: 'uint64',
  int32: 'int32',
  uint32: 'uint32',
  int16: 'int64',
  uint16: 'uint64',
  int8: 'int64',
  uint8: 'uint64',
};

const stridedOutCtor: Partial<
  Record<DType, new (buf: ArrayBuffer, off: number, len: number) => TypedArray>
> = {
  float64: Float64Array as unknown as new (
    buf: ArrayBuffer,
    off: number,
    len: number
  ) => TypedArray,
  float32: Float32Array as unknown as new (
    buf: ArrayBuffer,
    off: number,
    len: number
  ) => TypedArray,
  float16: Float32Array as unknown as new (
    buf: ArrayBuffer,
    off: number,
    len: number
  ) => TypedArray,
  int64: BigInt64Array as unknown as new (buf: ArrayBuffer, off: number, len: number) => TypedArray,
  uint64: BigUint64Array as unknown as new (
    buf: ArrayBuffer,
    off: number,
    len: number
  ) => TypedArray,
  int32: Int32Array as unknown as new (buf: ArrayBuffer, off: number, len: number) => TypedArray,
  uint32: Uint32Array as unknown as new (buf: ArrayBuffer, off: number, len: number) => TypedArray,
  int16: BigInt64Array as unknown as new (buf: ArrayBuffer, off: number, len: number) => TypedArray,
  uint16: BigUint64Array as unknown as new (
    buf: ArrayBuffer,
    off: number,
    len: number
  ) => TypedArray,
  int8: BigInt64Array as unknown as new (buf: ArrayBuffer, off: number, len: number) => TypedArray,
  uint8: BigUint64Array as unknown as new (
    buf: ArrayBuffer,
    off: number,
    len: number
  ) => TypedArray,
};

const stridedOutBpe: Partial<Record<DType, number>> = {
  float64: 8,
  float32: 4,
  float16: 4,
  int64: 8,
  uint64: 8,
  int32: 4,
  uint32: 4,
  int16: 8,
  uint16: 8, // output is i64/u64 = 8 bytes
  int8: 8,
  uint8: 8, // output is i64/u64 = 8 bytes
};

/**
 * WASM-accelerated strided prod along an axis.
 * Output dtype may be promoted (narrow ints → i64/u64).
 * Returns output ArrayStorage, or null if WASM can't handle.
 */
export function wasmReduceProdStrided(
  a: ArrayStorage,
  outerSize: number,
  axisSize: number,
  innerSize: number
): ArrayStorage | null {
  if (!a.isCContiguous) return null;

  const totalSize = outerSize * axisSize * innerSize;
  if (totalSize < BASE_THRESHOLD * wasmConfig.thresholdMultiplier) return null;

  const dtype = a.dtype;
  const kernel = stridedKernels[dtype];
  const InCtor = ctorMap[dtype];
  const OutCtor = stridedOutCtor[dtype];
  const outDtype = stridedOutDtype[dtype];
  const outBpe = stridedOutBpe[dtype];
  if (!kernel || !InCtor || !OutCtor || !outDtype || !outBpe) return null;

  const inBpe = (InCtor as unknown as { BYTES_PER_ELEMENT: number }).BYTES_PER_ELEMENT;
  const outSize = outerSize * innerSize;

  wasmConfig.wasmCallCount++;
  resetScratchAllocator();
  let inPtr: number;
  if (dtype === 'float16') {
    inPtr = f16InputToScratchF32(a, totalSize);
  } else {
    inPtr = resolveInputPtr(a.data, a.isWasmBacked, a.wasmPtr, a.offset, totalSize, inBpe);
  }
  const outPtr = alloc(outSize * outBpe);

  kernel(inPtr, outPtr, outerSize, axisSize, innerSize);

  const outData = copyOut(outPtr, outSize, OutCtor);

  // f16: cast f32 output back to f16 (matches NumPy overflow behavior)
  if (dtype === 'float16') {
    const f16Data = new Float16Array(outSize);
    f16Data.set(outData as Float32Array);
    return ArrayStorage.fromData(f16Data as unknown as TypedArray, [outSize], 'float16');
  }

  return ArrayStorage.fromData(outData, [outSize], outDtype);
}
